esp/platform/espprotocol.cpp

```diff
@@ -809,20 +811,36 @@ class ConnectionCloseDetector : implements IConnectionCloseDetector, public CInt
         if (m_callbackfunc)
             m_callbackfunc(m_callbackdata);
     }
+    bool checkClose()
+    {
+        int ret = m_sock->wait_read(0);
+        if (ret < 0)
+            return true;
+        else if (ret > 0)
+        {
+            if (m_sock->avail_read() == 0)
+                return true;
+        }
+        return false;
+    }
 public:
     IMPLEMENT_IINTERFACE;
-    ConnectionCloseDetector(IEspContext* ctx)
+    ConnectionCloseDetector(IEspContext* ctx, CloseNotifyCallback callbackfunc, void* callbackdata) : m_callbackfunc(callbackfunc), m_callbackdata(callbackdata)
     {
         if (!ctx || !ctx->queryRequest())
             return;
         CHttpRequest* req = dynamic_cast<CHttpRequest*>(ctx->queryRequest());
         if (!req)
             return;
-        ISocket* sock = req->getSocket();
-        if (!sock)
+        m_sock = req->getSocket();
+        if (!m_sock)
```

Comment:
Is it ok for ctor to return without creating m_sock ?  How will parent know it failed ?

Reply:
I can throw an exception.